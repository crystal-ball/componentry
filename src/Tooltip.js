/* eslint-disable  react/prop-types */
import generateContainer from './toggleable/generate-container';
const Container = generateContainer('tooltip');

import { getTextWidth } from './utils/dom';

/**
 * The Tooltip component creates an expandable info container on hover.
 *
 * TODO:
 * - Check body overflow on render and adjust width if needed
 * @class Tooltip
 * @constructor
 * @extends React.Component
 */
export default class Tooltip extends Container {
  // Control arias used by subcomponents
  contentArias = { id: true, role: 'tooltip' }
  triggerArias = { describedby: true }

  /**
   * Internal cache for width of tooltip content. Set after calculating content
   * width and reused on subsequent renders if content text has not changed.
   */
  contentWidth = null;
  /**
   * Internal cache for tooltip content. Used to check if the content has changed
   * between showings of tooltip.
   */
  content = null;

  /**
   * Handle calling user hooks and toggling internal state if component is not
   * controlled.
   */
  toggleActive = e => {
    let { active } = this.props;
    // Handle uncontrolled tooltip with internal state
    if (active === undefined) { active = this.state.active; }

    if (!active) {
      // Tooltip is not active, so it is transitioning to shown
      this.props.onActivate(this, e);

      // Position absolute tooltip is constrained by the parent width. Set tooltip
      // width to content width to overflow parent bounds
      const contentElement = document.getElementById(this.guid);
      const content = contentElement.innerText;
      this.content = content;

      if (content === this.content && this.contentWidth) {
        // If width has already been calculated and content has not changed, use
        // cached width for performance
        contentElement.style.width = `${this.contentWidth}px`;
      } else {
        // Get all styles of content element, set width and cache
        const styles = window.getComputedStyle(contentElement);
        // Get padding, font size and font family of content
        const width = getTextWidth(content, `${styles.fontSize} ${styles.fontFamily}`)
          + parseFloat(styles.paddingLeft)
          + parseFloat(styles.paddingRight)
          + 1;

        contentElement.style.width = `${width}px`;
        this.contentWidth = width;
      }
    } else {
      // Tooltip is active, so it is transitioning to hidden
      this.props.onDeactivate(this, e);
    }

    if (this.props.active === undefined) {
      // State is not controlled, update internal state
      this.setState({ active: !active });
    }

    if (!active) {
      this.props.onActivated(this, e);
    } else {
      this.props.onDeactivated(this, e);
    }
  }
}
