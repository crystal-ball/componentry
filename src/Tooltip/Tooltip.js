import activeContainer from '../factories/active-container-component'
import activeContent from '../factories/active-content-component'
import activeTrigger from '../factories/active-trigger-component'

/**
 * [Tooltip component 📝](https://componentry.design/components/tooltip)
 */
const Tooltip = activeContainer('tooltip', {
  escEvents: true,
  mouseEvents: true,
})
export default Tooltip

/**
 * [Tooltip trigger component 📝](https://componentry.design/components/tooltip)
 */
Tooltip.Content = activeContent('tooltip', {
  arias: { id: true, role: 'tooltip', hidden: true },
  positioned: true,
})

/**
 * [Tooltip content component 📝](https://componentry.design/components/tooltip)
 */
Tooltip.Trigger = activeTrigger('tooltip', {
  arias: { describedby: true },
})
