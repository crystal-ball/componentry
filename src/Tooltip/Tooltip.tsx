import { activeContainerBuilder } from '../utils/active-container-component-builder'
import { activeActionBuilder } from '../utils/active-action-component-builder'
import { activeContentBuilder } from '../utils/active-content-component-builder'
import {
  ActiveActionBaseProps,
  ActiveContainerBaseProps,
  ActiveContentBaseProps,
  ComponentBaseProps,
} from '../utils/types'

export interface TooltipProps
  extends ActiveContainerBaseProps,
    ComponentBaseProps<'div'> {}

export interface TooltipActionProps
  extends ActiveActionBaseProps,
    ComponentBaseProps<'button'> {
  /** Display variant */
  variant?: 'primary'
}

export interface TooltipContentProps
  extends ActiveContentBaseProps,
    ComponentBaseProps<'div'> {
  /** Display variant */
  variant?: 'primary'
}

export interface Tooltip {
  (props: TooltipProps): React.ReactElement
  /**
   * [Tooltip action component üìù](https://componentry.design/components/tooltip)
   */
  Action: React.FC<TooltipActionProps>
  /**
   * [Tooltip content component üìù](https://componentry.design/components/tooltip)
   */
  Content: React.FC<TooltipContentProps>
}

/**
 * [Tooltip component üìù](https://componentry.design/components/tooltip)
 */
export const Tooltip = activeContainerBuilder<TooltipProps>('Tooltip', {
  escEvents: true,
  mouseEvents: true,
}) as Tooltip

Tooltip.Action = activeActionBuilder<TooltipActionProps>('TooltipAction', {
  aria: { describedby: true },
})

Tooltip.Content = activeContentBuilder<TooltipContentProps>('TooltipContent', {
  aria: { id: true, role: 'tooltip', hidden: true },
  positioned: true,
})
