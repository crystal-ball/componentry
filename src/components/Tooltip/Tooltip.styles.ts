import { Theme } from '../../theme/theme'

//                                       <Tooltip /> styles
// --------------------------------------------------------

const tooltipArrowWidth = 10 // in pixels

export const tooltipStyles = (theme: Theme) => ({
  '.C9Y-Tooltip-base': {
    display: 'inline-block',
    position: 'relative',
  },

  // --- ACTION
  '.C9Y-TooltipAction': {},

  // --- CONTENT POSITIONER
  '.C9Y-TooltipContent': {
    // Content container overrides the width constraints for parent element
    width: '300px',
    position: 'absolute',
    opacity: 0,
    transition: 'opacity 0.2s cubic-bezier(0.4, 0, 0.2, 1)',

    '&.C9Y-active': {
      opacity: 1,
    },
  },

  // --- CONTENT
  '.C9Y-TooltipContentContents': {
    backgroundColor: theme.colors.gray[800],
    borderRadius: theme.borderRadius.DEFAULT,
    color: theme.colors.inverse,
    fontSize: theme.fontSize.sm,
    marginTop: `${tooltipArrowWidth}px`,
    maxWidth: '300px',
    padding: '0.25rem 0.5rem',
    position: 'relative',
    display: 'inline-block',
    textAlign: 'left',
    wordWrap: 'break-word', // Allow breaking very long words so they don't overflow the tooltip's bounds
  },

  // --- ARROW
  '.C9Y-TooltipContentArrow': {
    height: `${tooltipArrowWidth * 2}px`,
    left: '0.5rem',
    overflow: 'hidden',
    pointerEvents: 'none', // Prevents mouseenter of tip that slightly overlaps action
    position: 'absolute',
    top: `${tooltipArrowWidth * -1}px`,
    width: `${tooltipArrowWidth * 2}px`,

    '&:after': {
      background: theme.colors.gray[800],
      content: `''`,
      height: tooltipArrowWidth,
      left: `${tooltipArrowWidth / 2}px`,
      position: 'absolute',
      top: `${tooltipArrowWidth * 1.5}px`,
      transform: 'rotate(45deg)',
      width: tooltipArrowWidth,
    },
  },
})
