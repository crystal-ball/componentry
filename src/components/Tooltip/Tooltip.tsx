import React from 'react'
import { activeActionBuilder } from '../../utils/active-action-component-builder'
import { activeContainerBuilder } from '../../utils/active-container-component-builder'
import { activeContentBuilder } from '../../utils/active-content-component-builder'
import {
  ActiveActionBaseProps,
  ActiveContainerBaseProps,
  ActiveContentBaseProps,
} from '../../utils/base-types'
import { UtilityProps } from '../../utils/utility-classes'
import { Link } from '../Link/Link'

export interface TooltipProps
  extends ActiveContainerBaseProps,
    UtilityProps,
    React.ComponentPropsWithoutRef<'div'> {}

export interface TooltipActionProps
  extends ActiveActionBaseProps,
    UtilityProps,
    React.ComponentPropsWithoutRef<'button'> {
  /** Display variant */
  variant?: 'primary'
}

export interface TooltipContentProps
  extends ActiveContentBaseProps,
    UtilityProps,
    React.ComponentPropsWithoutRef<'div'> {
  /** Display variant */
  variant?: 'primary'
}

export interface Tooltip {
  (props: TooltipProps): React.ReactElement
  /**
   * [Tooltip action component üìù](https://componentry.design/components/tooltip)
   */
  Action: React.FC<TooltipActionProps>
  /**
   * [Tooltip content component üìù](https://componentry.design/components/tooltip)
   */
  Content: React.FC<TooltipContentProps>
}

/**
 * [Tooltip component üìù](https://componentry.design/components/tooltip)
 * @experimental
 */
export const Tooltip = activeContainerBuilder('Tooltip', {
  escEvents: true,
  mouseEvents: true,
}) as Tooltip

Tooltip.Action = activeActionBuilder('TooltipAction', {
  aria: { describedby: true },
  defaultAs: Link,
})

Tooltip.Content = activeContentBuilder('TooltipContent', {
  aria: { id: true, role: 'tooltip', hidden: true },
  defaultAs: TooltipContentElement,
})

function TooltipContentElement({
  children,
  renderArrow = true,
  ...rest
}: {
  children: React.ReactNode
  renderArrow: boolean
}) {
  return (
    <div {...rest}>
      {renderArrow && <div className='C9Y-TooltipContentArrow' />}
      <div className='C9Y-TooltipContentContents'>{children}</div>
    </div>
  )
}
