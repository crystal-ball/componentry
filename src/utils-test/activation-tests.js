/* eslint-disable import/no-extraneous-dependencies */
import React from 'react'
import { mount } from 'enzyme'

import dt from './dt'

/**
 * This set of activation event tests should pass for every element with active
 * state. It's probably not really necessary to run these tests for each element,
 * but this provides a simple high level gut check on if functionality is working
 * as expected for each component.
 *
 * IMPORTANT NOTES
 * - `dt()` is a shorthand method for creating data-test strings. Using data-test
 *   strings lets us reuse shared tests across components
 * - Enzyme `find()` will find **EVERY** instance of a prop. Since we are passing
 *   data-test attributes down here, it will always find more than one. Use
 *   `.last()` to selecte final node.
 */

const findContent = wrapper => wrapper.find(dt('content')).last()
const findTrigger = wrapper => wrapper.find(dt('trigger')).last()

export default TestComponent => {
  /**
   * Test that by default Active.Trigger elements render as button nodes, and
   * when passed an href render anchor nodes.
   */
  test('should render button or anchor tags by default', () => {
    const wrapper = mount(
      <TestComponent>
        <TestComponent.Trigger data-test="btn" />
        <TestComponent.Trigger data-test="anchor" href="http://localhost:3000/">
          Anchor
        </TestComponent.Trigger>
        <TestComponent.Content data-test="content" />
      </TestComponent>,
    )

    const btn = wrapper.find(dt('btn')).last()
    const anchor = wrapper.find(dt('anchor')).last()

    expect(anchor.type()).toEqual('a')
    expect(anchor.prop('type')).toEqual(undefined) // should not have type="button"
    expect(btn.type()).toEqual('button')
    expect(btn.prop('type')).toEqual('button') // should have type="button"
  })

  /**
   * Test that the default uncontrolled state scenario works. The active state should
   * be managed internally if not overriden with props and clicking the trigger should
   * show/hide content.
   */
  test('should update arias when trigger is activated', () => {
    const wrapper = mount(
      <TestComponent>
        <TestComponent.Trigger data-test="trigger" />
        <TestComponent.Content data-test="content" />
      </TestComponent>,
    )

    // Content
    const content = findContent(wrapper)
    const guid = content.prop('id')
    expect(content.length).toEqual(1)
    expect(content.prop('aria-hidden')).toEqual('true')

    // Trigger
    const trigger = findTrigger(wrapper)
    expect(trigger.length).toEqual(1)

    // TODO: need to figure out common aria attributes and test only those here
    // eg Popover and tooltip do not have aria-controls attrs
    if (trigger.prop('aria-controls')) expect(trigger.prop('aria-controls')).toEqual(guid)

    // Simulate a click, arias should update
    trigger.simulate('click')

    expect(findContent(wrapper).prop('aria-hidden')).toEqual('false')

    // Simulate a click, arias should update
    trigger.simulate('click')

    expect(findContent(wrapper).prop('aria-hidden')).toEqual('true')
  })

  /**
   * Test that all of the user hooks are being called on activation/deactivation.
   * TODO: update to check that event hooks are called with correct args.
   */
  test('should call user event hooks', () => {
    const onActivate = jest.fn()
    const onActivated = jest.fn()
    const onDeactivate = jest.fn()
    const onDeactivated = jest.fn()
    const wrapper = mount(
      <TestComponent
        onActivate={onActivate}
        onActivated={onActivated}
        onDeactivate={onDeactivate}
        onDeactivated={onDeactivated}
      >
        <TestComponent.Trigger data-test="trigger" />
        <TestComponent.Content data-test="content" />
      </TestComponent>,
    )

    const trigger = findTrigger(wrapper)

    trigger.simulate('click')

    expect(onActivate).toHaveBeenCalledTimes(1)
    expect(onActivated).toHaveBeenCalledTimes(1)
    expect(onDeactivate).toHaveBeenCalledTimes(0) // not yet!
    expect(onDeactivated).toHaveBeenCalledTimes(0) // not yet!

    trigger.simulate('click')

    expect(onActivate).toHaveBeenCalledTimes(1) // only once
    expect(onActivated).toHaveBeenCalledTimes(1) // only once
    expect(onDeactivate).toHaveBeenCalledTimes(1)
    expect(onDeactivated).toHaveBeenCalledTimes(1)
  })

  /**
   * Test setting initial active state using defaultActive prop
   */
  test('state should default to defaultActive when passed', () => {
    const wrapper = mount(
      <TestComponent defaultActive>
        <TestComponent.Trigger data-test="trigger" />
        <TestComponent.Content data-test="content" />
      </TestComponent>,
    )

    expect(findContent(wrapper).prop('aria-hidden')).toEqual('false')
  })

  /**
   * State checks if active prop value is different from internal state and updates
   * if that's the case, but this should only happen when passed active is an actual
   * value and not undefined. Check this with an instance with no value, and a
   * forced rerender
   */
  test('should not use active prop when value is undefined', () => {
    const wrapper = mount(
      <TestComponent>
        <TestComponent.Trigger data-test="trigger" />
        <TestComponent.Content data-test="content" />
      </TestComponent>,
    )
    expect(findContent(wrapper).prop('aria-hidden')).toEqual('true')
    wrapper.update()
    expect(findContent(wrapper).prop('aria-hidden')).toEqual('true')
  })

  /**
   * Test controlled component for methods `activate` and `deactivate`. Passing these
   * should override internal activate/deactivate meaning that clicking the trigger
   * for an element should only call those passed props.
   */
  test('should use passed activate and deactivate functions', () => {
    const activate = jest.fn()
    const deactivate = jest.fn()
    const wrapper = mount(
      <TestComponent activate={activate} deactivate={deactivate}>
        <TestComponent.Trigger data-test="trigger" />
        <TestComponent.Content data-test="content" />
      </TestComponent>,
    )

    const trigger = findTrigger(wrapper)

    trigger.simulate('click')

    // We have overridden the activate event so, dropdown should still be closed
    // and clicking trigger should call activate again

    expect(activate).toHaveBeenCalledTimes(1)
    expect(deactivate).toHaveBeenCalledTimes(0)
    expect(findContent(wrapper).prop('aria-hidden')).toEqual('true')

    trigger.simulate('click')

    expect(activate).toHaveBeenCalledTimes(2)
    expect(deactivate).toHaveBeenCalledTimes(0)
    expect(findContent(wrapper).prop('aria-hidden')).toEqual('true')

    // Mock controlled component being passed true, click should now ONLY call deactivate
    wrapper.setProps({ active: true })
    trigger.simulate('click')

    expect(activate).toHaveBeenCalledTimes(2)
    expect(deactivate).toHaveBeenCalledTimes(1)
    expect(findContent(wrapper).prop('aria-hidden')).toEqual('false')

    trigger.simulate('click')

    expect(activate).toHaveBeenCalledTimes(2)
    expect(deactivate).toHaveBeenCalledTimes(2)
    expect(findContent(wrapper).prop('aria-hidden')).toEqual('false')
  })

  /**
   * Test that any flavor of the `<Active />` component exposes internal state and
   * state change methods using the FaCC pattern
   */
  test('should expose internals using FaCC pattern', () => {
    const wrapper = mount(
      <TestComponent>
        {({ active, activate, deactivate }) => (
          <div>
            <span data-test="active">{String(active)}</span>
            <button type="button" onClick={activate} data-test="activate" />
            <button type="button" onClick={deactivate} data-test="deactivate" />
          </div>
        )}
      </TestComponent>,
    )

    expect(wrapper.find(dt('active')).text()).toEqual('false')
    wrapper.find(dt('activate')).simulate('click')
    expect(wrapper.find(dt('active')).text()).toEqual('true')
  })
}
