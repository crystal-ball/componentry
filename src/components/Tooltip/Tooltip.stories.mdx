import { Meta } from '@storybook/addon-docs'
import { BADGE } from '@geometricpanda/storybook-addon-badges'

import { Tooltip } from './Tooltip'

<Meta
  title='Components/Tooltip'
  component={Tooltip}
  parameters={{ badges: [BADGE.EXPERIMENTAL] }}
/>

# Tooltip

Tooltip extends Active with support for mouse events and styling to match typical tooltip
titles.

<Tooltip
  onActivate={console.log}
  onActivated={console.log}
  onDeactivate={console.log}
  onDeactivated={console.log}
>
  <Tooltip.Action>Fun fact</Tooltip.Action>
  <Tooltip.Content>
    Only 8% of the worldâ€™s currency is physical money, the rest only exists on computers.
  </Tooltip.Content>
</Tooltip>
