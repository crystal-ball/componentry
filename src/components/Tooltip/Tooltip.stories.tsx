import type { Meta, StoryObj } from '@storybook/react'

import { Tooltip } from './Tooltip'

const meta: Meta<typeof Tooltip> = {
  component: Tooltip,
}

export default meta
type Story = StoryObj<typeof Tooltip>

export const Primary: Story = {
  render: () => (
    <Tooltip active>
      <Tooltip.Action>Fun fact</Tooltip.Action>
      <Tooltip.Content>
        Only 8% of the world's currency is physical money, the rest only exists on
        computers.
      </Tooltip.Content>
    </Tooltip>
  ),
}
