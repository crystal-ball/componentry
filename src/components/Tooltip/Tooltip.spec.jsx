import { render, screen } from '@testing-library/react'
import { describe, it, expect } from 'vitest'

import { activationTests } from '../../test/activation-tests'
import { elementTests } from '../../test/element-tests'
import { Tooltip } from './Tooltip'

describe('<Tooltip />', () => {
  // Basic library activation test suite
  activationTests(Tooltip, { name: 'Tooltip', testArias: ['describedby'] })
  // Basic library element test suite
  elementTests(Tooltip)
  elementTests(Tooltip.Action)
  elementTests(Tooltip.Content, { mounted: 'always' })
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Tooltip /> snapshots', () => {
  it('renders correctly', () => {
    render(
      <Tooltip data-testid='tooltip'>
        <Tooltip.Action>Action</Tooltip.Action>
        <Tooltip.Content mounted='always'>Content</Tooltip.Content>
      </Tooltip>,
    )

    expect(screen.getByTestId('tooltip')).toMatchSnapshot()
  })
})
