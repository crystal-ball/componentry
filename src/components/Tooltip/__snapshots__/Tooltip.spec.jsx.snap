// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Tooltip /> snapshots > renders correctly 1`] = `
<div
  class="C9Y-Tooltip-base"
  data-id="test-guid"
  data-testid="tooltip"
>
  <button
    aria-describedby="test-guid"
    class="C9Y-Link-base C9Y-Link-text C9Y-TooltipAction"
    type="button"
  >
    Action
  </button>
  <div
    aria-hidden="true"
    class="C9Y-TooltipContent"
    id="test-guid"
    role="tooltip"
  >
    <div
      class="C9Y-TooltipContentArrow"
    />
    <div
      class="C9Y-TooltipContentContents"
    >
      Content
    </div>
  </div>
</div>
`;
